printer/dot/printer.go

```diff
@@ -33,14 +42,55 @@ func (d Dot) Print(g *graph.Graph, w io.Writer) error {
 		if pv == nil {
 			pv = provider.RawProvider{}
 		}
-		shape := "ellipse"
+		attr := map[string]string{
+			"shape": "ellipse",
+		}
 		if pv.IsEdge(rs) {
-			shape = "rectangle"
+			attr["shape"] = "rectangle"
+		}
+
+		if opt.ShowIcons && n.Resource.Icon != "" {
+			ext := path.Ext(n.Resource.Icon)
+			pngIcon := fmt.Sprintf("%s.png", n.Resource.Icon[0:len(n.Resource.Icon)-len(ext)])
+			assetPath := path.Join("inframap", "assets", pv.Type().String(), pngIcon)
+			pathIcon := path.Join(xdg.CacheHome, assetPath)
+
+			attr["image"] = fmt.Sprintf("%q", pathIcon)
+			attr["shape"] = "plaintext"
+			attr["labelloc"] = "b"
+			attr["height"] = "1.15"
+
+			// If the file does not exists on the Cache path, we have to write it,
+			// if not it means it's already correct so nothing to be done
+			if _, err := os.Stat(pathIcon); os.IsNotExist(err) {
+				p, err := xdg.CacheFile(assetPath)
+				if err != nil {
+					return err
+				}
+
+				f, err := os.Create(p)
+				if err != nil {
+					return err
+				}
+
+				iconFile, err := pkger.Open(path.Join("/assets", "icons", pv.Type().String(), pngIcon))
+				if err != nil {
+					return err
+				}
+
+				_, err = io.Copy(f, iconFile)
+				if err != nil {
+					return err
+				}
+
+				f.Close()
+				iconFile.Close()
```

Comment:
same as `f`, we can move it right after its creation and a `defer` statement

Reply:
Same I mentioned before, having a `defer` inside a `for` loop is not recommended.